### PR TITLE
Lexer improvement

### PR DESCRIPTION
**Pull request by SebastianMestre**
I moved the awful lexer stuff into a data structure.

It's horrendously inefficient in terms of memory usage, and probably a lot slower in runtime than before. I still think it's a win just due to the maintainability gains.

(edit)

You can think about this as an incomplete table based lexer. It might be a good idea to move to a full one eventually, but meh, baby steps.

### CODE REVIEW

File: src/lexer.hpp [hunk 1]
```diff
@@ -52,4 +48,15 @@ struct Lexer {
...
+	std::vector<char> m_source;
+	TokenArray& m_tokens;
+
+	Trie m_symbols_trie;
+
+	int m_source_index {0};
+	int m_token_index {0};
+
+	int m_current_line {0};
+	int m_current_column {0};
```
Comment (lushoBarlett) [inline src/lexer.hpp:61]: Why did you move this down here?

File: src/lexer.cpp [hunk 2]
```diff
@@ -102,244 +148,84 @@ std::pair<bool, TokenTag> Lexer::is_keyword(InternedString const& str) {
...
-		case '=':
-			push_token(TokenTag::DECLARE_ASSIGN, 2);
-			break;
-		default:
-			push_token(TokenTag::DECLARE, 1);
-		}
-		break;
+	char const* base_ptr = m_source.data() + i0;
+	m_tokens.push_back(
+			{TokenTag::STRING, InternedString {base_ptr, len}, l0, c0, m_current_line, m_current_column});
```
Comment (lushoBarlett) [inline src/lexer.cpp:182]: This seems indented weirdly.

File: src/algorithms/trie.hpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		int data;
+	};
+
+	struct Node {
+		Node();
+		Node(Node&&) = default;
+		Node(Node const&) = default;
+
+		uint16_t entry {nil};
+		std::array<uint16_t, 255> children;
```
Comment (lushoBarlett) [inline src/algorithms/trie.hpp:22]: Well this is... some memory usage.

File: src/algorithms/trie.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
+#include "trie.hpp"
+#include <iostream>
+#include <cassert>
+
+Trie::Node::Node() {
+	children.fill(nil);
+}
+
+Trie::Trie (){
```
Comment (lushoBarlett) [inline src/algorithms/trie.cpp:9]: Open bracket inconsistency

File: src/algorithms/trie.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+void Trie::insert(Entry entry) {
+	int entry_idx = entries.size();
+	entries.push_back(entry);
+
+	int n = 0;
+	for (char c : entry.text) {
+		uint8_t uc = c;
+		auto nn = go(n, uc);
```
Comment (lushoBarlett) [inline src/algorithms/trie.cpp:24]: Perhaps more purpose indicative identifiers would be nice here.

Comment (lushoBarlett) [reply src/algorithms/trie.cpp:24]: And everywhere for that matter.

**Review by lushoBarlett — APPROVED** [review]
I looked at it a bunch, struggled a bit to understand, but ultimately found no bugs. I do agree to merge it, but I'm not happy with it as is.

The good thing about it at least is that for our purposes it's not super slow, because we use it for symbols and we don't use that many. The Lexer was never a bottleneck to begin with.

I do notice how you are leaning more towards abstraction instead of optimization now for some reason with the Lexer. I see it as a good balance to have, perhaps we did some very premature optimizations there that were not necessary yet.

File: src/lexer.hpp (hunk 1 above)
Comment (lushoBarlett) [reply src/lexer.hpp:61]: I draw attention to this because it is inconsistent with the rest of the codebase.

Comment (SebastianMestre) [reply src/lexer.hpp:61]: I think we should move to that style everywhere.

When you look at a class definition, you more often want to know what the methods are. It's just more convenient to read.

File: src/lexer.cpp (hunk 2 above)
Comment (SebastianMestre) [reply src/lexer.cpp:182]: Yeah I just cut and pasted without worrying about it

File: src/algorithms/trie.cpp (hunk 5 above)
Comment (SebastianMestre) [reply src/algorithms/trie.cpp:24]: Lol it's very competitive programming-y

Comment (SebastianMestre) [reply src/algorithms/trie.cpp:24]: FWIW, c stands for char, uc for unsigned char, n for node, and nn for next node.

I'll change the names

**Comment by SebastianMestre** [comment]
> I do notice how you are leaning more towards abstraction instead of optimization now for some reason with the Lexer. I see it as a good balance to have, perhaps we did some very premature optimizations there that were not necessary yet.

Meh, it was not optimized before, really. It was written in the most straightforward way possible, which happened to be *decently* fast (it was probably not branch prediction friendly, though).

I want to write a really fast lexer eventually, too, though.
